pkg/dockerregistry/server/pullthroughblobstore.go

```diff
@@ -116,30 +118,33 @@ func (r *pullthroughBlobStore) proxyStat(ctx context.Context, retriever importer
 }
 
 // ServeBlob attempts to serve the requested digest onto w, using a remote proxy store if necessary.
-func (r *pullthroughBlobStore) ServeBlob(ctx context.Context, w http.ResponseWriter, req *http.Request, dgst digest.Digest) error {
-	store, ok := r.digestToStore[dgst.String()]
+func (pbs *pullthroughBlobStore) ServeBlob(ctx context.Context, w http.ResponseWriter, req *http.Request, dgst digest.Digest) error {
+	store, ok := pbs.digestToStore[dgst.String()]
 	if !ok {
-		return r.BlobStore.ServeBlob(ctx, w, req, dgst)
+		return pbs.BlobStore.ServeBlob(ctx, w, req, dgst)
 	}
 
-	desc, err := store.Stat(ctx, dgst)
-	if err != nil {
-		context.GetLogger(ctx).Errorf("failed to stat digest %q: %v", dgst.String(), err)
-		return err
-	}
+	// store the content locally if requested, but ensure only one instance at a time
+	// is storing to avoid excessive local writes
+	if pbs.mirror {
+		mu.Lock()
+		if _, ok = inflight[dgst]; ok {
+			mu.Unlock()
+			_, err := pbs.copyContent(store, ctx, dgst, w, req)
+			return err
+		}
+		inflight[dgst] = struct{}{}
+		mu.Unlock()
 
-	remoteReader, err := store.Open(ctx, dgst)
-	if err != nil {
-		context.GetLogger(ctx).Errorf("failure to open remote store for digest %q: %v", dgst.String(), err)
-		return err
+		go func(dgst digest.Digest) {
```

Comment:
Could we utilize [io.Multiwriter](https://golang.org/pkg/io/#MultiWriter) to avoid reading the downloading the blob twice?